controller/aks-cluster-config-handler.go

```diff
@@ -498,386 +466,132 @@ func BuildUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.Secr
-	if authorizer == nil {
-		return nil, fmt.Errorf("authorizer must not be nil")
-	}
-
-	baseURL := spec.BaseURL
-	if baseURL != nil {
-		baseURL = &azure.PublicCloud.ResourceManagerEndpoint
-	}
-
-	client := containersv209.NewContainerServicesClientWithBaseURI(*baseURL, spec.SubscriptionID)
-	client.Authorizer = authorizer
-
-	return &client, nil
-}
-
-// createOrUpdateCluster creates a new managed Kubernetes cluster
-func createOrUpdateCluster(ctx context.Context, secretsCache wranglerv1.SecretCache, clusterClient *containersv211.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (c containersv211.ManagedCluster, err error) {
-
-	dnsPrefix := spec.DNSPrefix
-	if dnsPrefix == nil {
-		dnsPrefix = to.StringPtr(spec.ClusterName)
-	}
-
-	tags := make(map[string]*string)
-	for key, val := range spec.Tags {
-		if val != "" {
-			tags[key] = to.StringPtr(val)
-		}
-	}
-	displayName := spec.ClusterName
-	if displayName == "" {
-		displayName = spec.ClusterName
+		return err
 	}
-	tags["displayName"] = to.StringPtr(displayName)
 
-	kubernetesVersion := spec.KubernetesVersion
-
-	var vmNetSubnetID *string
-	networkProfile := &containersv211.NetworkProfile{}
-	if hasCustomVirtualNetwork(spec) {
-		virtualNetworkResourceGroup := spec.ResourceGroup
-
-		//if virtual network resource group is set, use it, otherwise assume it is the same as the cluster
-		if spec.VirtualNetworkResourceGroup != nil {
-			virtualNetworkResourceGroup = *spec.VirtualNetworkResourceGroup
-		}
-
-		vmNetSubnetID = to.StringPtr(fmt.Sprintf(
-			"/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v/subnets/%v",
-			spec.SubscriptionID,
-			virtualNetworkResourceGroup,
-			spec.VirtualNetwork,
-			spec.Subnet,
-		))
-
-		networkProfile.DNSServiceIP = spec.NetworkDNSServiceIP
-		networkProfile.DockerBridgeCidr = spec.NetworkDockerBridgeCIDR
-		networkProfile.ServiceCidr = spec.NetworkServiceCIDR
-
-		if spec.NetworkPlugin != nil {
-			networkProfile.NetworkPlugin = containersv211.Kubenet
+	updateNodePool := false
+	// check for updated NodePools
+	upstreamNodePools := utils.BuildNodePoolMap(upstreamSpec.NodePools)
+	for _, np := range spec.NodePools {
+		upstreamNodePool, ok := upstreamNodePools[to.String(np.Name)]
+		if ok {
+			// There is a matching node pool in the cluster already, so update it if needed
+			if to.Int32(np.Count) != to.Int32(upstreamNodePool.Count) {
+				logrus.Infof("Updating node count in node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+				updateNodePool = true
+			}
+			if np.EnableAutoScaling != nil && to.Bool(np.EnableAutoScaling) != to.Bool(upstreamNodePool.EnableAutoScaling) {
+				logrus.Infof("Updating autoscaling in node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+				updateNodePool = true
+			}
+			if np.OrchestratorVersion != nil && to.String(np.OrchestratorVersion) != to.String(upstreamNodePool.OrchestratorVersion) {
```

Comment:
nit: it seems like the nil check was re-added though it's not needed https://github.com/rancher/aks-operator/pull/3#discussion_r621702549

Reply:
I would like to keep it as it is.